src/ui/components/pages/importEmail/signContext.tsx

```diff
@@ -0,0 +1,161 @@
+  }, []);
+
+  const onCancel = useCallback(() => {
+    setModalOpen(false);
+    modalPromiseRef.current.resolve?.(false);
+    modalPromiseRef.current.resolve = null;
+  }, []);
+
+  const contextValue = useMemo(
+    () => ({ openConfirmModal, closeConfirmModal }),
+    [closeConfirmModal, openConfirmModal]
+  );
+
+  return (
+    <>
+      <SignContext.Provider value={contextValue}>
+        {children}
+      </SignContext.Provider>
+
+      {account?.type === 'wx' && (
+        <Modal showModal={modalOpen} animation={Modal.ANIMATION.FLASH}>
+          <div className={clsx('modal', 'cover', styles.root)}>
+            <div className={styles.content}>
+              <Button
+                className="modal-close"
+                onClick={onCancel}
+                type="button"
+                view="transparent"
+              />
+
+              <h2 className={clsx('margin4', 'title1')}>
+                {t('importEmail.loginRequired')}
+              </h2>
+
+              <Login
+                userData={{ username: account.username, password: '' }}
+                onConfirm={onOk}
+              />
+            </div>
+          </div>
+        </Modal>
+      )}
+
+      {account?.type === 'ledger' && (
+        <Modal animation={Modal.ANIMATION.FLASH} showModal={modalOpen}>
+          <LedgerConnectModal
+            networkCode={account.networkCode}
+            onClose={onCancel}
+            onReady={onOk}
+          />
+        </Modal>
+      )}
+    </>
+  );
+};
+
+export function useSign<OnConfirmParams>(
+  onConfirm: (params: OnConfirmParams) => void | Promise<void>
+): [(params: OnConfirmParams) => Promise<void>, boolean] {
+  const account = usePopupSelector(state => state.selectedAccount);
+
+  const [pending, setPending] = useState(false);
+
+  const { openConfirmModal, closeConfirmModal } = useContext(SignContext);
+
+  const sign = useCallback(
+    async (params: OnConfirmParams) => {
+      setPending(true);
+
+      switch (account?.type) {
+        case 'wx':
+          try {
+            await Background.identityRestore(account.uuid);
+            onConfirm(params);
+          } catch (e) {
+            if (await openConfirmModal()) {
+              await Background.identityUpdate();
+              onConfirm(params);
+              closeConfirmModal();
+            }
+          }
+          break;
+        case 'ledger':
+          await ledgerService.updateStatus(account.networkCode);
+
+          if (ledgerService.status === LedgerServiceStatus.Ready) {
+            onConfirm(params);
+          } else {
+            if (await openConfirmModal()) {
+              onConfirm(params);
+              closeConfirmModal();
+            }
+          }
+          break;
+        default:
+          onConfirm(params);
+          break;
+      }
+
+      setPending(false);
+    },
+    [account, onConfirm, openConfirmModal, closeConfirmModal]
+  );
+
+  return [sign, pending];
```

Comment:
I think it makes sense to use an object instead of a tuple here, so we decide about names right here. I see no reason to allow giving these values arbitrary names at a call site. It's not a generic thing, like e.g. `useState`